File: server-pedidos-java/src/main/java/br/edu/utfpr/serverpedidosjava/controller/dto/ClienteDTO.java
package br.edu.utfpr.serverpedidosjava.controller.dto;

import br.edu.utfpr.serverpedidosjava.entity.Cliente;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import org.hibernate.validator.constraints.br.CPF;

public record ClienteDTO(
        Integer id,
        @NotBlank(message = "{nome.notblank}")
        String nome,
        @NotBlank(message = "{cpf.notblank}")
        @CPF(message = "{cpf.invalido}")
        String cpf,
        @NotBlank(message = "{telefone.notblank}")
        String telefone,
        @NotNull(message = "{endereco.notnull}")
        @Valid
        EnderecoDTO endereco
) {
    public Cliente toEntity() {
        Cliente cliente = new Cliente();
        cliente.setId(this.id);
        cliente.setNome(this.nome);
        cliente.setCpf(this.cpf);
        cliente.setTelefone(this.telefone);
        cliente.setEndereco(this.endereco.toEntity(cliente));
        return cliente;
    }

    public static ClienteDTO fromEntity(Cliente cliente) {
        return new ClienteDTO(
                cliente.getId(),
                cliente.getNome(),
                cliente.getCpf(),
                cliente.getTelefone(),
                EnderecoDTO.fromEntity(cliente.getEndereco())
        );
    }
}

File: server-pedidos/src/main/kotlin/br/edu/utfpr/serverpedidos/controller/dto/ClienteDTO.kt
package br.edu.utfpr.serverpedidos.controller.dto

import br.edu.utfpr.serverpedidos.entity.Cliente
import br.edu.utfpr.serverpedidos.util.validator.Telefone
import jakarta.validation.Valid
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull
import org.hibernate.validator.constraints.br.CPF

data class ClienteDTO(
    val id: Int = 0,
    @field:NotBlank(message = "{nome.notblank}")
    val nome: String = "",
    @field:NotBlank(message = "{cpf.notblank}")
    @field:CPF(message = "{cpf.invalido}")
    val cpf: String = "",
    @field:NotBlank(message = "{telefone.notblank}")
    @field:Telefone
    val telefone: String = "",
    @field:NotNull(message = "{endereco.notnull}")
    @field:Valid
    val endereco: EnderecoDTO? = null
) {
    fun toEntity(): Cliente = Cliente(
        id = this.id,
        nome = this.nome,
        cpf = this.cpf,
        telefone = this.telefone
    ).also {
        it.endereco = endereco?.toEntity(it)
    }

    companion object {
        fun fromEntity(cliente: Cliente): ClienteDTO = ClienteDTO(
            id = cliente.id,
            nome = cliente.nome,
            cpf = cliente.cpf,
            telefone = cliente.telefone,
            endereco = cliente.endereco?.let { EnderecoDTO.fromEntity(it) }
        )
    }
}
